uber/models.py

```diff
@@ -1523,16 +1543,44 @@ def weighted_hours(self):
     def total_hours(self):
         return self.weighted_hours * self.slots
 
-    @cached_property
-    def all_staffers(self):
-        return self.session.query(Attendee).order_by(Attendee.last_first).all()
+    def _potential_staff(self, staff_only=False, order_by=Attendee.full_name):
+        """
+        return a list of attendees who:
+        1) are assigned to this job's location
+        2) are allowed to work this job (job is unrestricted, or they're trusted in this job's location)
 
-    @cached_property
-    def available_staffers(self):
-        return [s for s in self.all_staffers
-                if self.location in s.assigned_depts_ints
-                   and (s.trusted or not self.restricted)
-                   and self.no_overlap(s)]
+        :param: staff_only: restrict result to attendees where staffing==True
+        :param: order_by: order by another Attendee attribute
+        """
+        return (self.session.query(Attendee)
+                .filter(Attendee.assigned_depts.contains(str(self.location)))
+                .filter(*[Attendee.trusted_depts.contains(str(self.location))] if self.restricted else [])
+                .filter_by(**{'staffing': True} if staff_only else {})
+                .order_by(order_by)
+                .all())
+
+    @property
+    def capable_staff_opts(self):
+        # format output for use with the {% options %} template decorator
+        return [(a.id, a.full_name) for a in self.capable_staff]
+
+    @property
+    def capable_staff(self):
+        """
+        Return a list of staffers who could sign up for this job.
+
+        Important: Just because a staffer is capable of working
+        this job doesn't mean they are available to work.
+        They may have other shift hours during that time period.
+        """
+        return self._potential_staff(staff_only=True)
+
+    @property
+    def available_staff(self):
+        """
+        Return a list of staffers who are allowed to sign up for this Job and have the free time
+        """
+        return [s for s in self._potential_staff(order_by=Attendee.last_first) if self.no_overlap(s)]
 
```

Comment:
OK. definitely pay attention to this part, this is all new stuff.

I took the remains of the 'staff_by_job' query, and the 'Job.available_staffers' and distilled them into the following two methods:
1) capable_staff - people who could work this job and may or may not be free
2) available_staff - people who could work this job and are in fact free

They use a common new function, _potential_staff(), because they're doing the exact same filtering (i.e. is an attendee trusted in this location / job unrestricted / are they assigned).

I was meticulous with the unit tests during my refactor to make sure these returned the exact same output.

I did remove the ```@cached_property``` on the original 'available_staff' property.  I am guessing this was on there because that function would return every single  attendee in the database and then filter like 95% of them in python.  the new version of this function does all that work in the DB which should be a hell of a lot faster.  Eli, let me know if you agree that we probably don't need ```@cached_property``` anymore.

That being said, on closer examination, there is probably more stuff to distill out of here:

1) _potential_staff has the option to filter out attendees where Attendee.staffing is not True.  I left this option in because I wasn't sure if there was ever a way to have an attendee able to sign up for a shift when they weren't marked as staffing.  While this is the original functionality, I bet.... it's probably just a thing the original query forgot to check for.  Since this would have been a rare edge case the result is the same.  I think it would be safe to remove staff_only=True here and just filter by attendee.staffing=True all the time.

2) Ordering: One of these functions wants to order by full name, the other by last.first name.  I think this is probably something that doesn't belong in these properties, and should probably be moved into the UI side of things.  Not too worried about it for now and it preserves the functionality.

This was one hell of a refactor : )

The unit test side of this (below) is also a pretty big changeup. Check that out too.